File: mobile/lib/features/activity/dm_resurface.dart
import 'feed_item.dart';
import '../../shared/relay/relay.dart';

final _hexPubkey = RegExp(r'^[0-9a-f]{64}$');

Set<String> _peerPubkeys({
  required String authorPubkey,
  required List<List<String>> tags,
  required String currentPubkey,
}) {
  final self = currentPubkey.trim().toLowerCase();
  final candidates = <String>{authorPubkey.trim().toLowerCase()};
  for (final tag in tags) {
    if (tag.length >= 2 && tag[0] == 'p') {
      candidates.add(tag[1].trim().toLowerCase());
    }
  }
  candidates.removeWhere(
    (pubkey) => pubkey == self || !_hexPubkey.hasMatch(pubkey),
  );
  return candidates;
}

Set<String> dmPeerPubkeysFromFeedItem(FeedItem item, String currentPubkey) =>
    _peerPubkeys(
      authorPubkey: item.pubkey,
      tags: item.tags,
      currentPubkey: currentPubkey,
    );

Set<String> dmPeerPubkeysFromEvent(NostrEvent event, String currentPubkey) =>
    _peerPubkeys(
      authorPubkey: event.pubkey,
      tags: event.tags,
      currentPubkey: currentPubkey,
    );

bool isIncomingDmMessageEvent(NostrEvent event, String currentPubkey) {
  final self = currentPubkey.trim().toLowerCase();
  return event.channelId != null &&
      EventKind.channelMessageEventKinds.contains(event.kind) &&
      event.pubkey.toLowerCase() != self &&
      event.tags.any(
        (tag) =>
            tag.length >= 2 && tag[0] == 'p' && tag[1].toLowerCase() == self,
      );
}

File: desktop/src/features/channels/dmResurface.ts
import type {
  FeedItem,
  HomeFeedResponse,
  RelayEvent,
} from "@/shared/api/types";
import { CHANNEL_MESSAGE_EVENT_KINDS } from "@/shared/constants/kinds";
import { normalizePubkey } from "@/shared/lib/pubkey";

const CHANNEL_MESSAGE_KINDS = new Set<number>(CHANNEL_MESSAGE_EVENT_KINDS);
const HEX_PUBKEY = /^[0-9a-f]{64}$/;

export function dmPeerPubkeysFromFeedItem(
  item: FeedItem,
  currentPubkey: string | undefined,
): string[] {
  const self = normalizePubkey(currentPubkey ?? "");
  const candidates = [
    item.pubkey,
    ...item.tags
      .filter((tag) => tag[0] === "p" && typeof tag[1] === "string")
      .map((tag) => tag[1]),
  ];

  return [...new Set(candidates.map(normalizePubkey))].filter(
    (pubkey) => HEX_PUBKEY.test(pubkey) && pubkey !== self,
  );
}

export function dmPeerPubkeysFromRelayEvent(
  event: RelayEvent,
  currentPubkey: string | undefined,
): string[] {
  return dmPeerPubkeysFromFeedItem(
    {
      id: event.id,
      kind: event.kind,
      pubkey: event.pubkey,
      content: event.content,
      createdAt: event.created_at,
      channelId:
        event.tags.find((tag) => tag[0] === "h" && tag[1])?.[1] ?? null,
      channelName: "",
      tags: event.tags,
      category: "mention",
    },
    currentPubkey,
  );
}

export function isIncomingDmMessageFeedItem(
  item: FeedItem,
  currentPubkey: string | undefined,
): boolean {
  const self = normalizePubkey(currentPubkey ?? "");
  if (
    !item.channelId ||
    self.length === 0 ||
    !CHANNEL_MESSAGE_KINDS.has(item.kind) ||
    normalizePubkey(item.pubkey) === self
  ) {
    return false;
  }

  return item.tags.some(
    (tag) => tag[0] === "p" && normalizePubkey(tag[1] ?? "") === self,
  );
}

export function isIncomingDmMessageRelayEvent(
  event: RelayEvent,
  currentPubkey: string | undefined,
): boolean {
  return isIncomingDmMessageFeedItem(
    {
      id: event.id,
      kind: event.kind,
      pubkey: event.pubkey,
      content: event.content,
      createdAt: event.created_at,
      channelId:
        event.tags.find((tag) => tag[0] === "h" && tag[1])?.[1] ?? null,
      channelName: "",
      tags: event.tags,
      category: "mention",
    },
    currentPubkey,
  );
}

export function markHiddenDmFeedItems(
  feed: HomeFeedResponse,
  hiddenDmIds: ReadonlySet<string>,
): HomeFeedResponse {
  if (hiddenDmIds.size === 0) return feed;

  const mark = (item: FeedItem): FeedItem =>
    item.channelId && hiddenDmIds.has(item.channelId)
      ? { ...item, channelType: "dm" }
      : item;

  return {
    ...feed,
    feed: {
      mentions: feed.feed.mentions.map(mark),
      needsAction: feed.feed.needsAction.map(mark),
      activity: feed.feed.activity.map(mark),
      agentActivity: feed.feed.agentActivity.map(mark),
    },
  };
}
